File: lib/mascotUnlock.js
import { getBenchmarkForProfile, getSwimLevel } from './swimBenchmarks.js';
import { getMonthlyChallengeHistory, getMonthKey } from './swimMonthlyChallenges.js';

const MASCOT_IDS = ['flip', 'flo', 'fins'];

const getStatsSessions = (sessions) => (sessions || []).filter((session) => !session?.excludeFromStats);

const average = (values) => {
  if (!values.length) return null;
  return values.reduce((sum, value) => sum + value, 0) / values.length;
};

const PACE_LEVEL_RANK = {
  unknown: 0,
  developing: 0,
  beginner: 1,
  intermediate: 2,
  advanced: 3,
};

/** Unlock requirements: reach pace level OR earn enough monthly tier medals. */
export const MASCOT_UNLOCK_REQUIREMENTS = {
  flip: { minPaceLevel: null, minMonthlyMedals: 0 },
  flo: { minPaceLevel: 'intermediate', minMonthlyMedals: 5 },
  fins: { minPaceLevel: 'advanced', minMonthlyMedals: 10 },
};

export function getUserSwimPaceLevel(profile, sessions) {
  if (!profile?.sex || !profile?.age) return 'unknown';

  const statsSessions = getStatsSessions(sessions);
  if (!statsSessions.length) return 'unknown';

  const benchmark = getBenchmarkForProfile(profile.sex, profile.age);
  const paces = statsSessions
    .map((session) => session.metrics?.paceSecPer100m)
    .filter((pace) => pace != null);
  const pace = average(paces)
    ?? statsSessions[statsSessions.length - 1]?.metrics?.paceSecPer100m
    ?? null;

  if (pace == null) return 'unknown';
  return getSwimLevel(pace, benchmark);
}

export function countMonthlyMedals(sessions, monthlyChallengeRerolls = {}) {
  return getMonthlyChallengeHistory(sessions, { monthlyChallengeRerolls })
    .filter((entry) => entry.tier)
    .length;
}

export function meetsPaceRequirement(paceLevel, requiredLevel) {
  if (!requiredLevel) return true;
  return (PACE_LEVEL_RANK[paceLevel] || 0) >= (PACE_LEVEL_RANK[requiredLevel] || 0);
}

export function getMascotUnlockStatus(mascotId, { profile, sessions, monthlyChallengeRerolls = {} } = {}) {
  const requirements = MASCOT_UNLOCK_REQUIREMENTS[mascotId];
  if (!requirements) {
    return { unlocked: false, paceMet: false, medalsMet: false, paceLevel: 'unknown', monthlyMedals: 0 };
  }

  if (!requirements.minPaceLevel && !requirements.minMonthlyMedals) {
    return { unlocked: true, paceMet: true, medalsMet: true, paceLevel: 'beginner', monthlyMedals: 0 };
  }

  const paceLevel = getUserSwimPaceLevel(profile, sessions);
  const monthlyMedals = countMonthlyMedals(sessions, monthlyChallengeRerolls);
  const paceMet = meetsPaceRequirement(paceLevel, requirements.minPaceLevel);
  const medalsMet = monthlyMedals >= requirements.minMonthlyMedals;

  return {
    unlocked: paceMet || medalsMet,
    paceMet,
    medalsMet,
    paceLevel,
    monthlyMedals,
    requirements,
  };
}

export function isMascotUnlocked(mascotId, context) {
  return getMascotUnlockStatus(mascotId, context).unlocked;
}

export function getUnlockedMascotIds(context) {
  return MASCOT_IDS.filter((id) => isMascotUnlocked(id, context));
}

export function hasSessionsInMonth(sessions, monthKey = getMonthKey()) {
  return (sessions || []).some((session) => session.date?.startsWith(monthKey));
}

/**
 * Mascot switches are allowed once per calendar month, before the first swim
 * of that month (when new monthly goals are set).
 */
export function canSwitchMascot({
  profile,
  sessions,
  monthKey = getMonthKey(),
  nextMascotId,
  currentMascotId,
} = {}) {
  if (!nextMascotId || !MASCOT_IDS.includes(nextMascotId)) {
    return { allowed: false, reason: 'invalid' };
  }

  if (nextMascotId === currentMascotId) {
    return { allowed: true, reason: 'same' };
  }

  if (!isMascotUnlocked(nextMascotId, { profile, sessions })) {
    return { allowed: false, reason: 'locked' };
  }

  if (profile?.mascotSwitchMonthKey === monthKey) {
    return { allowed: false, reason: 'alreadySwitched' };
  }

  if (hasSessionsInMonth(sessions, monthKey)) {
    return { allowed: false, reason: 'afterFirstSession' };
  }

  return { allowed: true, reason: 'ok' };
}

export function resolveUnlockedMascotId(profile, context = {}) {
  const requested = profile?.mascotId;
  if (requested && MASCOT_IDS.includes(requested) && isMascotUnlocked(requested, context)) {
    return requested;
  }

  const legacy = profile?.mascotSex || profile?.sex;
  if (!requested && legacy === 'female' && isMascotUnlocked('flo', context)) {
    return 'flo';
  }

  return 'flip';
}

/** Resolve active mascot with unlock validation. */
export function resolveMascotId(profile, context = {}) {
  return resolveUnlockedMascotId(profile, {
    profile,
    sessions: context.sessions || [],
    monthlyChallengeRerolls: context.monthlyChallengeRerolls || {},
  });
}

File: ios/AapSC/Lib/MascotUnlock.swift
import Foundation

enum MascotUnlock {
    static let unlockRequirements: [String: (minPaceLevel: String?, minMonthlyMedals: Int)] = [
        "flip": (nil, 0),
        "flo": ("intermediate", 5),
        "fins": ("advanced", 10)
    ]

    private static let paceLevelRank: [String: Int] = [
        "unknown": 0,
        "developing": 0,
        "beginner": 1,
        "intermediate": 2,
        "advanced": 3
    ]

    static func userSwimPaceLevel(profile: SwimProfile, sessions: [SwimSession]) -> SwimLevel {
        guard !profile.sex.isEmpty, profile.age > 0 else { return .unknown }
        let stats = SwimAnalysis.statsSessions(sessions)
        guard !stats.isEmpty else { return .unknown }

        let benchmark = SwimBenchmarks.benchmark(for: profile.sex, age: profile.age)
        let paces = stats.compactMap(\.metrics.paceSecPer100m).filter { $0 > 0 }
        let pace: Int?
        if paces.isEmpty {
            pace = stats.last?.metrics.paceSecPer100m
        } else {
            pace = paces.reduce(0, +) / paces.count
        }
        guard let pace else { return .unknown }
        return SwimBenchmarks.swimLevel(paceSecPer100m: pace, benchmark: benchmark)
    }

    static func countMonthlyMedals(
        sessions: [SwimSession],
        monthlyChallengeRerolls: [String: MonthRerollEntry] = [:]
    ) -> Int {
        SwimMonthlyChallenges.getMonthlyChallengeHistory(
            sessions: sessions,
            monthlyChallengeRerolls: monthlyChallengeRerolls
        ).filter { $0.tier != nil }.count
    }

    static func meetsPaceRequirement(paceLevel: SwimLevel, requiredLevel: String?) -> Bool {
        guard let requiredLevel else { return true }
        let current = paceLevelRank[paceLevel.rawValue] ?? 0
        let required = paceLevelRank[requiredLevel] ?? 0
        return current >= required
    }

    static func unlockStatus(
        mascotId: String,
        profile: SwimProfile,
        sessions: [SwimSession],
        monthlyChallengeRerolls: [String: MonthRerollEntry] = [:]
    ) -> (unlocked: Bool, paceMet: Bool, medalsMet: Bool, paceLevel: SwimLevel, monthlyMedals: Int) {
        guard let requirements = unlockRequirements[mascotId] else {
            return (false, false, false, .unknown, 0)
        }

        if requirements.minPaceLevel == nil && requirements.minMonthlyMedals == 0 {
            return (true, true, true, .beginner, 0)
        }

        let paceLevel = userSwimPaceLevel(profile: profile, sessions: sessions)
        let monthlyMedals = countMonthlyMedals(
            sessions: sessions,
            monthlyChallengeRerolls: monthlyChallengeRerolls
        )
        let paceMet = meetsPaceRequirement(paceLevel: paceLevel, requiredLevel: requirements.minPaceLevel)
        let medalsMet = monthlyMedals >= requirements.minMonthlyMedals

        return (paceMet || medalsMet, paceMet, medalsMet, paceLevel, monthlyMedals)
    }

    static func isUnlocked(
        mascotId: String,
        profile: SwimProfile,
        sessions: [SwimSession],
        monthlyChallengeRerolls: [String: MonthRerollEntry] = [:]
    ) -> Bool {
        unlockStatus(
            mascotId: mascotId,
            profile: profile,
            sessions: sessions,
            monthlyChallengeRerolls: monthlyChallengeRerolls
        ).unlocked
    }

    static func resolveMascotId(
        profile: SwimProfile,
        sessions: [SwimSession],
        monthlyChallengeRerolls: [String: MonthRerollEntry] = [:]
    ) -> String {
        if let requested = profile.mascotId,
           MascotConstants.ids.contains(requested),
           isUnlocked(mascotId: requested, profile: profile, sessions: sessions, monthlyChallengeRerolls: monthlyChallengeRerolls) {
            return requested
        }
        return "flip"
    }

    static func hasSessionsInMonth(_ sessions: [SwimSession], monthKey: String) -> Bool {
        sessions.contains { $0.date.hasPrefix(monthKey) }
    }

    static func canSwitchMascot(
        profile: SwimProfile,
        sessions: [SwimSession],
        monthKey: String = SwimMonthlyChallenges.getMonthKey(),
        nextMascotId: String,
        currentMascotId: String
    ) -> MascotSwitchResult {
        guard MascotConstants.ids.contains(nextMascotId) else {
            return MascotSwitchResult(allowed: false, reason: "invalid")
        }
        if nextMascotId == currentMascotId {
            return MascotSwitchResult(allowed: true, reason: "same")
        }
        if !isUnlocked(mascotId: nextMascotId, profile: profile, sessions: sessions) {
            return MascotSwitchResult(allowed: false, reason: "locked")
        }
        if profile.mascotSwitchMonthKey == monthKey {
            return MascotSwitchResult(allowed: false, reason: "alreadySwitched")
        }
        if hasSessionsInMonth(sessions, monthKey: monthKey) {
            return MascotSwitchResult(allowed: false, reason: "afterFirstSession")
        }
        return MascotSwitchResult(allowed: true, reason: "ok")
    }
}
